internal/db/lint/lint.go

```diff
@@ -0,0 +1,174 @@
+package lint
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/spf13/afero"
+	"github.com/spf13/viper"
+	"github.com/supabase/cli/internal/debug"
+	"github.com/supabase/cli/internal/utils"
+)
+
+var (
+	AllowedLevels = []string{
+		"warning",
+		"error",
+	}
+	//go:embed templates/check.sql
+	checkSchemaScript string
+)
+
+type LintLevel int
+
+func toEnum(level string) LintLevel {
+	for i, curr := range AllowedLevels {
+		if curr == level {
+			return LintLevel(i)
+		}
+	}
+	return -1
+}
+
+func Run(ctx context.Context, schema []string, level string, stdout io.Writer, fsys afero.Fs, opts ...func(*pgx.ConnConfig)) error {
+	// Sanity checks.
+	if err := utils.LoadConfigFS(fsys); err != nil {
+		return err
+	}
+	if err := utils.AssertSupabaseDbIsRunning(); err != nil {
+		return err
+	}
+	// Run lint script
+	conn, err := ConnectLocalPostgres(ctx, "localhost", utils.Config.Db.Port, "postgres", opts...)
+	if err != nil {
+		return err
+	}
+	defer conn.Close(ctx)
+	result, err := LintDatabase(ctx, conn, schema)
+	if err != nil {
+		return err
+	}
+	filtered := filterResult(result, toEnum(level))
+	// Print output
+	if len(filtered) == 0 {
+		return nil
+	}
+	enc := json.NewEncoder(stdout)
+	enc.SetIndent("", "  ")
+	if err := enc.Encode(filtered); err != nil {
+		return err
+	}
+	return nil
+}
+
+func filterResult(result []Result, minLevel LintLevel) (filtered []Result) {
+	for _, r := range result {
+		out := Result{Function: r.Function}
+		for _, issue := range r.Issues {
+			if toEnum(issue.Level) >= minLevel {
+				out.Issues = append(out.Issues, issue)
+			}
+		}
+		if len(out.Issues) > 0 {
+			filtered = append(filtered, out)
+		}
+	}
+	return filtered
+}
+
+// Connnect to local Postgres with optimised settings. The caller is responsible for closing the connection returned.
+func ConnectLocalPostgres(ctx context.Context, host string, port uint, database string, options ...func(*pgx.ConnConfig)) (*pgx.Conn, error) {
+	url := fmt.Sprintf("postgresql://postgres:postgres@%s:%d/%s", host, port, database)
+	// Parse connection url
+	config, err := pgx.ParseConfig(url)
+	if err != nil {
+		return nil, err
+	}
+	// Apply config overrides
+	for _, op := range options {
+		op(config)
+	}
+	if viper.GetBool("DEBUG") {
+		debug.SetupPGX(config)
+	}
+	// Connect to database
+	return pgx.ConnectConfig(ctx, config)
+}
+
+func LintDatabase(ctx context.Context, conn *pgx.Conn, schema []string) ([]Result, error) {
+	// Enable plpgsql_check
+	enable := "CREATE EXTENSION IF NOT EXISTS plpgsql_check"
+	if _, err := conn.Exec(ctx, enable); err != nil {
+		return nil, err
+	}
```

Comment:
This command should have no side effects, i.e. we should either:
- fail if `plpgsql_check` is not enabled, or
- enable it if not enabled, but disable it at the end (but don't disable it if it's already enabled)

Reply:
Or if there's no need to commit, we can simply rollback the txn

Reply:
Is there any benefit for disabling this extension on local dev? I'm thinking to add it to default schema after this PR.

Reply:
Same reason for disabling pgtap: the schema will diverge from the hosted platform's

Reply:
Sure, also added a task to include extensions in schema diff.